File: frontend/src/lib/onboarding.ts
import { apiRequest } from "@/lib/api";
import { getAuthToken } from "@/lib/auth";

export type OnboardingPayload = {
  current_band: string | null;
  target_band: string | null;
  study_hours_per_day: number | null;
  weakest_skill: string | null;
  strongest_skill: string | null;
  exam_type: string | null;
  purpose: string | null;
  country: string | null;
  exam_date: string | null;
};

export type OnboardingResponse = {
  id: number;
  current_band: string | null;
  target_band: string | null;
  study_hours_per_day: number | null;
  weakest_skill: string | null;
  strongest_skill: string | null;
  exam_type: string | null;
  purpose: string | null;
  country: string | null;
  exam_date: string | null;
};

export async function saveOnboarding ( payload: OnboardingPayload ) {
  const token = getAuthToken();

  if ( !token ) {
    throw new Error( "You must be logged in to save onboarding." );
  }

  return apiRequest<OnboardingResponse>( "/onboarding/save", {
    method: "POST",
    body: payload,
    token,
  } );
}

export async function getMyOnboarding () {
  const token = getAuthToken();

  if ( !token ) {
    return null;
  }

  return apiRequest<OnboardingResponse | null>( "/onboarding/me", {
    method: "GET",
    token,
  } );
}

File: backend/app/schemas/onboarding.py
from datetime import date
from typing import Optional

from pydantic import BaseModel, Field


class OnboardingRequest(BaseModel):
    current_band: Optional[str] = Field(default=None)
    target_band: Optional[str] = Field(default=None)

    study_hours_per_day: Optional[int] = Field(default=None)

    weakest_skill: Optional[str] = Field(default=None)
    strongest_skill: Optional[str] = Field(default=None)

    exam_type: Optional[str] = Field(default=None)

    purpose: Optional[str] = Field(default=None)

    country: Optional[str] = Field(default=None)

    exam_date: Optional[date] = Field(default=None)


class OnboardingResponse(BaseModel):
    id: int

    current_band: Optional[str]
    target_band: Optional[str]

    study_hours_per_day: Optional[int]

    weakest_skill: Optional[str]
    strongest_skill: Optional[str]

    exam_type: Optional[str]

    purpose: Optional[str]

    country: Optional[str]

    exam_date: Optional[date]

    class Config:
        from_attributes = True